### PR TITLE
[bluebird] Properly typed promisifyAll

### PR DESCRIPTION
**Pull request by agg23**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: http://bluebirdjs.com/docs/api/promise.promisifyall.html

Properly types `promisifyAll` using the new template literal support.

### CODE REVIEW

File: types/bluebird/index.d.ts
Comment (lhecker) [inline types/bluebird/index.d.ts:50]: It'd be nice if we could prevent `ExtractAsyncMethod` from returning non-never results in the first place. For instance...

This assumes that you removed the `extends any[]` constraints from `WithoutLast` and `Last`. I'm not sure if we need them. If you feel like we should keep them I'm fine with that.

We can introduce a new mapper method:
```ts
type PromiseMethod<A, R> = R extends never ? never : (...arg: WithoutLast<A>) => Promise<R>;
```

And use it:
```ts
type ExtractAsyncMethod<T> = T extends (...args: infer A) => any
    ? PromiseMethod<A, ExtractCallbackValueType<Last<A>>[0]>
    : never;
```

Afterwards we won't have to filter out invalid results using `ValidPromiseOrNever` anymore.
This ensures that we only have a single way to signal that a promisification failed: By returning `never`. Those are then centrally filtered out in `NonNeverKeys`/`NonNeverValues`.

The type names and implementation above are just an example of course.

---

By the way: I couldn't figure out why you use `Last<Required<A>>`. In my experiments `Last<A>` works just as well.

**Review by lhecker — CHANGES_REQUESTED** [review]
One more thing:
As much as it pains me personally, the type definitions for Bluebird use an indentation width of 2 spaces (the tests use 4). Please change your new code from 4 to 2 spaces within the index.d.ts file.

---

Please let me know what you think of my comments. They aren't necessarily "requirements" or anything for you to "fulfill" before merging this PR. Your generics are fairly complex and I'm also open for a discussion. I'd be happy if we could simplify them before merging this, but it's not strictly necessary to do it the way I suggested. 👍

File: types/bluebird/index.d.ts
Comment (agg23) [reply types/bluebird/index.d.ts:60]: Not quite, but combined with your `NonNeverValues` change, it serves the purpose.

File: types/bluebird/index.d.ts
Comment (agg23) [reply types/bluebird/index.d.ts:50]: Without `Required<T>`, functions that have an optional argument for the callback will not be present in the final result set. Tuples now store optional information:

```
type SomeObject = {
  optionalCallback: (value: string, callback?: (error: Error, output: number) => void) => void
}

Parameters<SomeObject['optionalCallback']>
/// [value: number, callback?: ((error: Error, value: string) => void) | undefined]
```

**Comment by agg23** [comment]
Thanks for all of the great feedback. I think all of your comments were addressed.

**Comment by lhecker** [comment]
@weswigham
This PR is fairly heavy on template logic and while I believe it's correct I don't _quite_ trust myself to be appropriate for approving it.
Could you (or some other maintainer) additionally review the new bits in [`types/bluebird/index.d.ts`](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52907/files#diff-9b548923c484c1fffb9ed8ca6023ee5b1b26ab662e05c6d22d46d7a3beef8920)?
I already checked the other parts of this PR and they are all correct to the extent of my knowledge.

**Pull request merged**
